File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/VB/UtilitiesLibrary/MeshLoader/XMeshManager.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.


Imports Microsoft.VisualBasic
Imports System
Imports Microsoft.WindowsAPICodePack.DirectX
Imports Microsoft.WindowsAPICodePack.DirectX.Direct3D
Imports Microsoft.WindowsAPICodePack.DirectX.Direct3D10
Imports Microsoft.WindowsAPICodePack.DirectX.Graphics
Imports System.Windows.Media.Media3D
Imports System.Runtime.InteropServices
Imports System.Reflection
Imports System.IO

Namespace Microsoft.WindowsAPICodePack.DirectX.DirectXUtilities

	''' <summary>
	''' Manages the XMesh file loading
	''' </summary>
	Public Class XMeshManager
		Implements IDisposable
		Friend device As D3DDevice

		Friend effect As Effect
		Friend techniqueRenderTexture As EffectTechnique
		Friend techniqueRenderVertexColor As EffectTechnique
		Friend techniqueRenderMaterialColor As EffectTechnique

		Friend brightnessVariable As EffectScalarVariable
		Friend materialColorVariable As EffectVectorVariable
		Friend worldVariable As EffectMatrixVariable
		Friend viewVariable As EffectMatrixVariable
		Friend projectionVariable As EffectMatrixVariable
		Friend diffuseVariable As EffectShaderResourceVariable

		''' <summary>
		''' Creates the mesh manager
		''' </summary>
		''' <param name="device"></param>
		Public Sub New(ByVal device As D3DDevice)
			Me.device = device

			' Create the effect
			'XMesh.fxo was compiled from XMesh.fx using:
			' "$(DXSDK_DIR)utilities\bin\x86\fxc" "$(ProjectDir)Mesh\MeshLoaders\XMesh.fx" /T fx_4_0 /Fo"$(ProjectDir)Mesh\MeshLoaders\XMesh.fxo"
            Using effectStream As Stream = System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("XMesh.fxo")
                effect = device.CreateEffectFromCompiledBinary(New BinaryReader(effectStream))
            End Using

			' Obtain the techniques
			techniqueRenderTexture = effect.GetTechniqueByName("RenderTextured")
			techniqueRenderVertexColor = effect.GetTechniqueByName("RenderVertexColor")
			techniqueRenderMaterialColor = effect.GetTechniqueByName("RenderMaterialColor")

			' Obtain the variables
			brightnessVariable = effect.GetVariableByName("Brightness").AsScalar()
			materialColorVariable = effect.GetVariableByName("MaterialColor").AsVector()
			worldVariable = effect.GetVariableByName("World").AsMatrix()
			viewVariable = effect.GetVariableByName("View").AsMatrix()
			projectionVariable = effect.GetVariableByName("Projection").AsMatrix()
			diffuseVariable = effect.GetVariableByName("tex2D").AsShaderResource()
		End Sub

		Public Sub SetViewAndProjection(ByVal view As Matrix4x4F, ByVal projection As Matrix4x4F)
            viewVariable.Matrix = view
            projectionVariable.Matrix = projection
		End Sub

		''' <summary>
		''' Returns an XMesh object that contains the data from a specified .X file.
		''' </summary>
		''' <param name="path"></param>
		''' <returns></returns>
		Public Function Open(ByVal path As String) As XMesh
			Dim mesh As New XMesh()
			mesh.Load(path, Me)
			Return mesh
		End Function

		''' <summary>
		''' Reutrns a specialization of an XMesh object that contains the data from a specified .X file
		''' </summary>
		''' <typeparam name="T"></typeparam>
		''' <param name="path"></param>
		''' <returns></returns>
		Public Function Open(Of T As {XMesh, New})(ByVal path As String) As T
			Dim mesh As New T()
			mesh.Load(path, Me)
			Return mesh
		End Function

		#Region "IDisposable Members"

		''' <summary>
		''' Cleans up the memory allocated by the manager.
		''' </summary>
		Public Sub Dispose() Implements IDisposable.Dispose
			effect.Dispose()
		End Sub

		#End Region
	End Class
End Namespace

File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/CS/UtilitiesLibrary/MeshLoader/XMeshManager.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using System;
using Microsoft.WindowsAPICodePack.DirectX;
using Microsoft.WindowsAPICodePack.DirectX.Direct3D;
using Microsoft.WindowsAPICodePack.DirectX.Direct3D10;
using Microsoft.WindowsAPICodePack.DirectX.Graphics;
using System.Windows.Media.Media3D;
using System.Runtime.InteropServices;
using System.Reflection;
using System.IO;

namespace Microsoft.WindowsAPICodePack.DirectX.DirectXUtilities
{

    /// <summary>
    /// Manages the XMesh file loading
    /// </summary>
    public class XMeshManager : IDisposable
    {
        internal D3DDevice device;

        internal Effect effect;
        internal EffectTechnique techniqueRenderTexture;
        internal EffectTechnique techniqueRenderVertexColor;
        internal EffectTechnique techniqueRenderMaterialColor;

        internal EffectScalarVariable brightnessVariable;
        internal EffectVectorVariable materialColorVariable;
        internal EffectMatrixVariable worldVariable;
        internal EffectMatrixVariable viewVariable;
        internal EffectMatrixVariable projectionVariable;
        internal EffectShaderResourceVariable diffuseVariable;

        /// <summary>
        /// Creates the mesh manager
        /// </summary>
        /// <param name="device"></param>
        public XMeshManager(D3DDevice device)
        {
            this.device = device;

            // Create the effect
            //XMesh.fxo was compiled from XMesh.fx using:
            // "$(DXSDK_DIR)utilities\bin\x86\fxc" "$(ProjectDir)Mesh\MeshLoaders\XMesh.fx" /T fx_4_0 /Fo"$(ProjectDir)Mesh\MeshLoaders\XMesh.fxo"
            using (Stream effectStream = Assembly.GetExecutingAssembly().GetManifestResourceStream(
                "Microsoft.WindowsAPICodePack.DirectX.DirectXUtilities.XMesh.fxo"))
            {
                effect = device.CreateEffectFromCompiledBinary( new BinaryReader( effectStream ) );
            }

            // Obtain the techniques
            techniqueRenderTexture = effect.GetTechniqueByName("RenderTextured");
            techniqueRenderVertexColor = effect.GetTechniqueByName("RenderVertexColor");
            techniqueRenderMaterialColor = effect.GetTechniqueByName("RenderMaterialColor");

            // Obtain the variables
            brightnessVariable = effect.GetVariableByName("Brightness").AsScalar;
            materialColorVariable = effect.GetVariableByName("MaterialColor").AsVector;
            worldVariable = effect.GetVariableByName("World").AsMatrix;
            viewVariable = effect.GetVariableByName("View").AsMatrix;
            projectionVariable = effect.GetVariableByName("Projection").AsMatrix;
            diffuseVariable = effect.GetVariableByName("tex2D").AsShaderResource;
        }

        public void SetViewAndProjection(Matrix4x4F view, Matrix4x4F projection)
        {
            viewVariable.Matrix = view;
            projectionVariable.Matrix = projection;
        }

        /// <summary>
        /// Returns an XMesh object that contains the data from a specified .X file.
        /// </summary>
        /// <param name="path"></param>
        /// <returns></returns>
        public XMesh Open(string path)
        {
            XMesh mesh = new XMesh();
            mesh.Load(path, this);
            return mesh;
        }

        /// <summary>
        /// Reutrns a specialization of an XMesh object that contains the data from a specified .X file
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="path"></param>
        /// <returns></returns>
        public T Open<T>(string path) where T : XMesh, new()
        {
            T mesh = new T();
            mesh.Load(path, this);
            return mesh;
        }

        #region IDisposable Members
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        private bool disposed;
        /// <summary>
        /// Cleans up the memory allocated by the manager.
        /// </summary>
        protected virtual void Dispose(bool disposing)
        {
            if (!disposed)
            {
                disposed = true;
                if (effect != null)
                    effect.Dispose();
                effect = null;

                if (techniqueRenderTexture != null)
                    techniqueRenderTexture.Dispose();
                techniqueRenderTexture = null;

                if (techniqueRenderVertexColor != null)
                    techniqueRenderVertexColor.Dispose();
                techniqueRenderVertexColor = null;

                if (techniqueRenderMaterialColor != null)
                    techniqueRenderMaterialColor.Dispose();
                techniqueRenderMaterialColor = null;

                if (brightnessVariable != null)
                    brightnessVariable.Dispose();
                brightnessVariable = null;

                if (materialColorVariable != null)
                    materialColorVariable.Dispose();
                materialColorVariable = null;

                if (worldVariable != null)
                    worldVariable.Dispose();
                worldVariable = null;

                if (viewVariable != null)
                    viewVariable.Dispose();
                viewVariable = null;

                if (projectionVariable != null)
                    projectionVariable.Dispose();
                projectionVariable = null;

                if (diffuseVariable != null)
                    diffuseVariable.Dispose();
                diffuseVariable = null;
            }
        }
        #endregion
    }
}
